### ISSUE TITLE
[patch] [resend] Turn $$ into a magical readonly variable that always fetches getpid() instead of caching it

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11179]
Migrated from [rt.perl.org#85520](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=85520) (status was 'resolved')

Searchable as RT85520$

**Comment by p5pRT** [comment]
### From @Corion

#### Created by @Corion




To get a Perlbug number\, I'm resending the patch appended
to this mail\.

commit df1cd6761ac9224f4aea7139509f9cacbf0b9ea0
Author&#8203;: Max Maischein \<corion@&#8203;corion\.net>
Date&#8203;:   Wed Dec 29 16&#8203;:35&#8203;:08 2010 \+0100

&nbsp;     Turn $$ into a magical readonly variable that always fetches 
getpid\(\) instead of caching it

&nbsp;     The intent is that by not caching $$\, we eliminate one opportunity 
for bugs&#8203;:
&nbsp;     If one embeds Perl or uses XS and calls fork\(3\) from C\, Perls 
notion of $$
&nbsp;     may go out of sync with what getpid\(\) returns\. By always fetching the
&nbsp;     value of $$ via getpid\(\)\, this bug opportunity is eliminated\. The 
overhead
&nbsp;     of always fetching $$ should be small and is likely only used for 
tempfile
&nbsp;     creation\, which should be dwarfed by file system accesses\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/gv.c b/gv.c
index 4e79171..f11f0cc 100644
--- a/gv.c
+++ b/gv.c
@@ -1457,6 +1457,8 @@ Perl_gv_fetchpvn_flags(pTHX_ const char *nambeg, 

```
</details>
STRLEN full_len, I32 flags,
&nbsp;  \#endif
&nbsp;          goto magicalize;

\+    case '$'&#8203;:        /\* $$ \*/
\+        goto magicalize;
&nbsp;      case '\!'&#8203;:        /\* $\! \*/
&nbsp;          GvMULTI\_on\(gv\);
&nbsp;          /\* If %\! has been used\, automatically load Errno\.pm\. \*/
<details><summary>Inline Patch</summary>

```diff
diff --git a/mg.c b/mg.c
index 8053bf1..89c22c6 100644
--- a/mg.c
+++ b/mg.c
@@ -1115,6 +1115,12 @@ Perl_magic_get(pTHX_ SV *sv, MAGIC *mg)
  	if (PL_ors_sv)
  	    sv_copypv(sv, PL_ors_sv);
  	break;
+    case '$': /* $$ */
+	sv_setiv(sv, (IV)PerlProc_getpid());
+	/* If you want another number, pull one at the entrance */
+	SvREADONLY_on(sv);
+	break;
+
      case '!':
  	{
  	dSAVE_ERRNO;
diff --git a/perl.c b/perl.c
index d2571a8..ab2e079 100644
--- a/perl.c
+++ b/perl.c
@@ -4125,11 +4125,6 @@ S_init_postdump_symbols(pTHX_ register int argc, 

```
</details>
register char **argv, register
&nbsp;  \#endif /\* \!PERL\_MICRO \*/
&nbsp;      \}
&nbsp;      TAINT\_NOT;
\-    if \(\(tmpgv = gv\_fetchpvs\("$"\, GV\_ADD|GV\_NOTQUAL\, SVt\_PV\)\)\) \{
\-        SvREADONLY\_off\(GvSV\(tmpgv\)\);
\-    sv\_setiv\(GvSV\(tmpgv\)\, \(IV\)PerlProc\_getpid\(\)\);
\-        SvREADONLY\_on\(GvSV\(tmpgv\)\);
\-    \}
&nbsp;  \#ifdef THREADS\_HAVE\_PIDS
&nbsp;      PL\_ppid = \(IV\)getppid\(\);
&nbsp;  \#endif
<details><summary>Inline Patch</summary>

```diff
diff --git a/pp_sys.c b/pp_sys.c
index 2845266..c4998bb 100644
--- a/pp_sys.c
+++ b/pp_sys.c
@@ -4084,12 +4084,6 @@ PP(pp_fork)
      if (childpid < 0)
  	RETSETUNDEF;
      if (!childpid) {
-	GV * const tmpgv = gv_fetchpvs("$", GV_ADD|GV_NOTQUAL, SVt_PV);
-	if (tmpgv) {
-            SvREADONLY_off(GvSV(tmpgv));
-	    sv_setiv(GvSV(tmpgv), (IV)PerlProc_getpid());
-            SvREADONLY_on(GvSV(tmpgv));
-        }
  #ifdef THREADS_HAVE_PIDS
  	PL_ppid = (IV)getppid();
  #endif
diff --git a/util.c b/util.c
index ac7dd57..b175c31 100644
--- a/util.c
+++ b/util.c
@@ -2770,12 +2770,6 @@ Perl_my_popen(pTHX_ const char *cmd, const char 

```
</details>
*mode)
&nbsp;        default\, binary\, low\-level mode; see PerlIOBuf\_open\(\)\. \*/
&nbsp;     PerlLIO\_setmode\(\(\*mode == 'r'\)\, O\_BINARY\);
&nbsp;  \#endif
\-
\-    if \(\(tmpgv = gv\_fetchpvs\("$"\, GV\_ADD|GV\_NOTQUAL\, SVt\_PV\)\)\) \{
\-        SvREADONLY\_off\(GvSV\(tmpgv\)\);
\-        sv\_setiv\(GvSV\(tmpgv\)\, PerlProc\_getpid\(\)\);
\-        SvREADONLY\_on\(GvSV\(tmpgv\)\);
\-    \}
&nbsp;  \#ifdef THREADS\_HAVE\_PIDS
&nbsp;      PL\_ppid = \(IV\)getppid\(\);
&nbsp;  \#endif
<details><summary>Inline Patch</summary>

```diff
diff --git a/win32/perlhost.h b/win32/perlhost.h
index 70a2f65..8103ee7 100644
--- a/win32/perlhost.h
+++ b/win32/perlhost.h
@@ -1722,18 +1722,11 @@ win32_start_child(LPVOID arg)
      PERL_SET_THX(my_perl);
      win32_checkTLS(my_perl);

-    /* set $$ to pseudo id */
  #ifdef PERL_SYNC_FORK
      w32_pseudo_id = id;
  #else
      w32_pseudo_id = GetCurrentThreadId();
  #endif
-    if (tmpgv = gv_fetchpv("$", TRUE, SVt_PV)) {
-	SV *sv = GvSV(tmpgv);
-	SvREADONLY_off(sv);
-	sv_setiv(sv, -(IV)w32_pseudo_id);
-	SvREADONLY_on(sv);
-    }
  #ifdef PERL_USES_PL_PIDSTATUS
      hv_clear(PL_pidstatus);
  #endif
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=low

Site configuration information for perl 5.13.10:

Configured by corion at Fri Mar  4 21:37:02 2011.

Summary of my perl5 (revision 5 version 13 subversion 10) configuration:
   Local Commit: de43a757c8705e286fca2f426d49cc107b810fd8
   Ancestor: 4d56cd4f546df82c1cabd288669bd8227d6847b4
   Platform:
     osname=MSWin32, osvers=5.1, archname=MSWin32-x86-multi-thread
     uname=''
     config_args='undef'
     hint=recommended, useposix=true, d_sigaction=undef
     useithreads=define, usemultiplicity=define
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=undef, use64bitall=undef, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='gcc', ccflags =' -s -O2 -DWIN32  -DPERL_IMPLICIT_CONTEXT 
-DPERL_IMPLICIT_SYS -fno-strict-aliasing -mms-bitfields',
     optimize='-s -O2',
     cppflags='-DWIN32'
     ccversion='', gccversion='4.4.3', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
     d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=12
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='long 
long', lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='g++', ldflags ='-s -L"c:\perl\lib\CORE" -L"C:\MinGW\lib"'
     libpth=C:\MinGW\lib
     libs=-lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 
-ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr 
-lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
     perllibs=-lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool 
-lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid 
-lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
     libc=, so=dll, useshrplib=true, libperl=libperl513.a
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags='-mdll -s -L"c:\perl\lib\CORE" 
-L"C:\MinGW\lib"'

Locally applied patches:



@INC for perl 5.13.10:
     C:/Projekte/bleadperl-git/lib
     .


Environment for perl 5.13.10:
     HOME (unset)
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
 
PATH=C:\strawberry\perl\bin;C:\strawberry\c\bin;C:\strawberry\perl\site\bin;C:\Programme\Microsoft 
IntelliType Pro\;C:\PROGRA~1\MATROX~1\System32;C:\Programme\MiKTeX 
2.7\miktex\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\Programme\Haufe\iDesk\iDeskService\;C:\Programme\Git\cmd;C:\Programme\GNU\GnuPG\pub;C:\Programme\sK1 
Project\UniConvertor-1.1.5\;C:\Programme\sK1 
Project\UniConvertor-1.1.5\DLLs;C:\Programme\Gemeinsame Dateien\GTK\2.0\bin
     PERL_BADLANG (unset)
     SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @exodist

I am indifferent about this patch\, but I want to point out that use of
$$ is not limited to temp file creation\. In current times
parallelization of tasks is important when possible\, specially with
multi\-core systems\. Currently there are several cpan modules which
allow you to run tasks in parallel\. When you do parallelization you
are almost always talking forked processes\. You need to determine if
you are the parent or a child\, if other forks have happened\, etc\. In
some of these systems such as Child and Parallel&#8203;::Runner $$ may be
called many times\. In the parallel testing framework Fennec forking
within tests is allowed\, but in order to accomplish this $$ must be
checked frequently\.

Once again I am indifferent to the patch\, I doubt the performance hit
will be meaningful anywhere\, but I wanted to point out that the
ramifications extend beyond temp file creation\.

\-Chad 'Exodist' Granum

On Sun\, Mar 6\, 2011 at 10&#8203;:20 AM\, Max Maischein
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Max Maischein
> \# Please include the string&#8203;:  \[perl \#85520\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=85520 >
>
>
> This is a bug report for perl from corion@&#8203;cpan\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.13\.10\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
>
> To get a Perlbug number\, I'm resending the patch appended
> to this mail\.
>
> commit df1cd6761ac9224f4aea7139509f9cacbf0b9ea0
> Author&#8203;: Max Maischein \<corion@&#8203;corion\.net>
> Date&#8203;:   Wed Dec 29 16&#8203;:35&#8203;:08 2010 \+0100
>
>     Turn $$ into a magical readonly variable that always fetches
> getpid\(\) instead of caching it
>
>     The intent is that by not caching $$\, we eliminate one opportunity
> for bugs&#8203;:
>     If one embeds Perl or uses XS and calls fork\(3\) from C\, Perls
> notion of $$
>     may go out of sync with what getpid\(\) returns\. By always fetching the
>     value of $$ via getpid\(\)\, this bug opportunity is eliminated\. The
> overhead
>     of always fetching $$ should be small and is likely only used for
> tempfile
>     creation\, which should be dwarfed by file system accesses\.
>
> diff \-\-git a/gv\.c b/gv\.c
> index 4e79171\.\.f11f0cc 100644
> \-\-\- a/gv\.c
> \+\+\+ b/gv\.c
> @&#8203;@&#8203; \-1457\,6 \+1457\,8 @&#8203;@&#8203; Perl\_gv\_fetchpvn\_flags\(pTHX\_ const char \*nambeg\,
> STRLEN full\_len\, I32 flags\,
>  \#endif
>            goto magicalize;
>
> \+       case '$'&#8203;:               /\* $$ \*/
> \+           goto magicalize;
>        case '\!'&#8203;:               /\* $\! \*/
>            GvMULTI\_on\(gv\);
>            /\* If %\! has been used\, automatically load Errno\.pm\. \*/
> diff \-\-git a/mg\.c b/mg\.c
> index 8053bf1\.\.89c22c6 100644
> \-\-\- a/mg\.c
> \+\+\+ b/mg\.c
> @&#8203;@&#8203; \-1115\,6 \+1115\,12 @&#8203;@&#8203; Perl\_magic\_get\(pTHX\_ SV \*sv\, MAGIC \*mg\)
>        if \(PL\_ors\_sv\)
>            sv\_copypv\(sv\, PL\_ors\_sv\);
>        break;
> \+    case '$'&#8203;: /\* $$ \*/
> \+       sv\_setiv\(sv\, \(IV\)PerlProc\_getpid\(\)\);
> \+       /\* If you want another number\, pull one at the entrance \*/
> \+       SvREADONLY\_on\(sv\);
> \+       break;
> \+
>      case '\!'&#8203;:
>        \{
>        dSAVE\_ERRNO;
> diff \-\-git a/perl\.c b/perl\.c
> index d2571a8\.\.ab2e079 100644
> \-\-\- a/perl\.c
> \+\+\+ b/perl\.c
> @&#8203;@&#8203; \-4125\,11 \+4125\,6 @&#8203;@&#8203; S\_init\_postdump\_symbols\(pTHX\_ register int argc\,
> register char \*\*argv\, register
>  \#endif /\* \!PERL\_MICRO \*/
>      \}
>      TAINT\_NOT;
> \-    if \(\(tmpgv = gv\_fetchpvs\("$"\, GV\_ADD|GV\_NOTQUAL\, SVt\_PV\)\)\) \{
> \-        SvREADONLY\_off\(GvSV\(tmpgv\)\);
> \-       sv\_setiv\(GvSV\(tmpgv\)\, \(IV\)PerlProc\_getpid\(\)\);
> \-        SvREADONLY\_on\(GvSV\(tmpgv\)\);
> \-    \}
>  \#ifdef THREADS\_HAVE\_PIDS
>      PL\_ppid = \(IV\)getppid\(\);
>  \#endif
> diff \-\-git a/pp\_sys\.c b/pp\_sys\.c
> index 2845266\.\.c4998bb 100644
> \-\-\- a/pp\_sys\.c
> \+\+\+ b/pp\_sys\.c
> @&#8203;@&#8203; \-4084\,12 \+4084\,6 @&#8203;@&#8203; PP\(pp\_fork\)
>      if \(childpid \< 0\)
>        RETSETUNDEF;
>      if \(\!childpid\) \{
> \-       GV \* const tmpgv = gv\_fetchpvs\("$"\, GV\_ADD|GV\_NOTQUAL\, SVt\_PV\);
> \-       if \(tmpgv\) \{
> \-            SvREADONLY\_off\(GvSV\(tmpgv\)\);
> \-           sv\_setiv\(GvSV\(tmpgv\)\, \(IV\)PerlProc\_getpid\(\)\);
> \-            SvREADONLY\_on\(GvSV\(tmpgv\)\);
> \-        \}
>  \#ifdef THREADS\_HAVE\_PIDS
>        PL\_ppid = \(IV\)getppid\(\);
>  \#endif
> diff \-\-git a/util\.c b/util\.c
> index ac7dd57\.\.b175c31 100644
> \-\-\- a/util\.c
> \+\+\+ b/util\.c
> @&#8203;@&#8203; \-2770\,12 \+2770\,6 @&#8203;@&#8203; Perl\_my\_popen\(pTHX\_ const char \*cmd\, const char
> \*mode\)
>        default\, binary\, low\-level mode; see PerlIOBuf\_open\(\)\. \*/
>     PerlLIO\_setmode\(\(\*mode == 'r'\)\, O\_BINARY\);
>  \#endif
> \-
> \-       if \(\(tmpgv = gv\_fetchpvs\("$"\, GV\_ADD|GV\_NOTQUAL\, SVt\_PV\)\)\) \{
> \-           SvREADONLY\_off\(GvSV\(tmpgv\)\);
> \-           sv\_setiv\(GvSV\(tmpgv\)\, PerlProc\_getpid\(\)\);
> \-           SvREADONLY\_on\(GvSV\(tmpgv\)\);
> \-       \}
>  \#ifdef THREADS\_HAVE\_PIDS
>        PL\_ppid = \(IV\)getppid\(\);
>  \#endif
> diff \-\-git a/win32/perlhost\.h b/win32/perlhost\.h
> index 70a2f65\.\.8103ee7 100644
> \-\-\- a/win32/perlhost\.h
> \+\+\+ b/win32/perlhost\.h
> @&#8203;@&#8203; \-1722\,18 \+1722\,11 @&#8203;@&#8203; win32\_start\_child\(LPVOID arg\)
>      PERL\_SET\_THX\(my\_perl\);
>      win32\_checkTLS\(my\_perl\);
>
> \-    /\* set $$ to pseudo id \*/
>  \#ifdef PERL\_SYNC\_FORK
>      w32\_pseudo\_id = id;
>  \#else
>      w32\_pseudo\_id = GetCurrentThreadId\(\);
>  \#endif
> \-    if \(tmpgv = gv\_fetchpv\("$"\, TRUE\, SVt\_PV\)\) \{
> \-       SV \*sv = GvSV\(tmpgv\);
> \-       SvREADONLY\_off\(sv\);
> \-       sv\_setiv\(sv\, \-\(IV\)w32\_pseudo\_id\);
> \-       SvREADONLY\_on\(sv\);
> \-    \}
>  \#ifdef PERL\_USES\_PL\_PIDSTATUS
>      hv\_clear\(PL\_pidstatus\);
>  \#endif
>
>
>
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=low
> \-\-\-
> Site configuration information for perl 5\.13\.10&#8203;:
>
> Configured by corion at Fri Mar  4 21&#8203;:37&#8203;:02 2011\.
>
> Summary of my perl5 \(revision 5 version 13 subversion 10\) configuration&#8203;:
>   Local Commit&#8203;: de43a757c8705e286fca2f426d49cc107b810fd8
>   Ancestor&#8203;: 4d56cd4f546df82c1cabd288669bd8227d6847b4
>   Platform&#8203;:
>     osname=MSWin32\, osvers=5\.1\, archname=MSWin32\-x86\-multi\-thread
>     uname=''
>     config\_args='undef'
>     hint=recommended\, useposix=true\, d\_sigaction=undef
>     useithreads=define\, usemultiplicity=define
>     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
>     use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
>     usemymalloc=n\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='gcc'\, ccflags =' \-s \-O2 \-DWIN32  \-DPERL\_IMPLICIT\_CONTEXT
> \-DPERL\_IMPLICIT\_SYS \-fno\-strict\-aliasing \-mms\-bitfields'\,
>     optimize='\-s \-O2'\,
>     cppflags='\-DWIN32'
>     ccversion=''\, gccversion='4\.4\.3'\, gccosandvers=''
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>     d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='long
> long'\, lseeksize=8
>     alignbytes=8\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='g\+\+'\, ldflags ='\-s \-L"c&#8203;:\\perl\\lib\\CORE" \-L"C&#8203;:\\MinGW\\lib"'
>     libpth=C&#8203;:\\MinGW\\lib
>     libs=\-lmoldname \-lkernel32 \-luser32 \-lgdi32 \-lwinspool \-lcomdlg32
> \-ladvapi32 \-lshell32 \-lole32 \-loleaut32 \-lnetapi32 \-luuid \-lws2\_32 \-lmpr
> \-lwinmm \-lversion \-lodbc32 \-lodbccp32 \-lcomctl32
>     perllibs=\-lmoldname \-lkernel32 \-luser32 \-lgdi32 \-lwinspool
> \-lcomdlg32 \-ladvapi32 \-lshell32 \-lole32 \-loleaut32 \-lnetapi32 \-luuid
> \-lws2\_32 \-lmpr \-lwinmm \-lversion \-lodbc32 \-lodbccp32 \-lcomctl32
>     libc=\, so=dll\, useshrplib=true\, libperl=libperl513\.a
>     gnulibc\_version=''
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
>     cccdlflags=' '\, lddlflags='\-mdll \-s \-L"c&#8203;:\\perl\\lib\\CORE"
> \-L"C&#8203;:\\MinGW\\lib"'
>
> Locally applied patches&#8203;:
>
>
> \-\-\-
> @&#8203;INC for perl 5\.13\.10&#8203;:
>     C&#8203;:/Projekte/bleadperl\-git/lib
>     \.
>
> \-\-\-
> Environment for perl 5\.13\.10&#8203;:
>     HOME \(unset\)
>     LANG \(unset\)
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>
> PATH=C&#8203;:\\strawberry\\perl\\bin;C&#8203;:\\strawberry\\c\\bin;C&#8203;:\\strawberry\\perl\\site\\bin;C&#8203;:\\Programme\\Microsoft
> IntelliType Pro\\;C&#8203;:\\PROGRA~1\\MATROX~1\\System32;C&#8203;:\\Programme\\MiKTeX
> 2\.7\\miktex\\bin;C&#8203;:\\WINDOWS\\system32;C&#8203;:\\WINDOWS;C&#8203;:\\WINDOWS\\System32\\Wbem;C&#8203;:\\Programme\\Haufe\\iDesk\\iDeskService\\;C&#8203;:\\Programme\\Git\\cmd;C&#8203;:\\Programme\\GNU\\GnuPG\\pub;C&#8203;:\\Programme\\sK1
> Project\\UniConvertor\-1\.1\.5\\;C&#8203;:\\Programme\\sK1
> Project\\UniConvertor\-1\.1\.5\\DLLs;C&#8203;:\\Programme\\Gemeinsame Dateien\\GTK\\2\.0\\bin
>     PERL\_BADLANG \(unset\)
>     SHELL \(unset\)
>
>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Mar 06 10&#8203;:20&#8203;:12 2011\, corion@&#8203;cpan\.org wrote&#8203;:
> To get a Perlbug number\, I'm resending the patch appended
> to this mail\.
> 
> commit df1cd6761ac9224f4aea7139509f9cacbf0b9ea0
> Author&#8203;: Max Maischein \<corion@&#8203;corion\.net>
> Date&#8203;:   Wed Dec 29 16&#8203;:35&#8203;:08 2010 \+0100
> 
>      Turn $$ into a magical readonly variable that always fetches
> getpid\(\) instead of caching it
> 
>      The intent is that by not caching $$\, we eliminate one
> opportunity
> for bugs&#8203;:
>      If one embeds Perl or uses XS and calls fork\(3\) from C\, Perls
> notion of $$
>      may go out of sync with what getpid\(\) returns\. By always fetching
> the
>      value of $$ via getpid\(\)\, this bug opportunity is eliminated\. The
> overhead
>      of always fetching $$ should be small and is likely only used for
> tempfile
>      creation\, which should be dwarfed by file system accesses\.

Is there any reason I should not apply this?

**Comment by p5pRT** [comment]
### From @cpansprout

Another message that RT didn’t pick up\, from
nntp&#8203;://nntp\.perl\.org/4DD979EB\.3090206@&#8203;cpan\.org&#8203;:

Hello\,

> On Sun Mar 06 10&#8203;:20&#8203;:12 2011\, corion@&#8203;cpan\.org wrote&#8203;:
>> To get a Perlbug number\, I'm resending the patch appended
>> to this mail\.
 >> \[\.\.\.\]
>>
> Is there any reason I should not apply this?
>

I'm unaware of any such reason\. I couldn't get a reply either for or 
against it from p5p at the time\, and also unfortunately didn't get the 
deprecation message into 5\.14\.

Still I think it should get applied\, as the two CPAN modules affected by 
it have patches\, and I think it fixes more latent bugs than it causes\.

\-max

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun May 22 18&#8203;:07&#8203;:52 2011\, sprout wrote&#8203;:
> Another message that RT didn’t pick up\, from
> nntp&#8203;://nntp\.perl\.org/4DD979EB\.3090206@&#8203;cpan\.org&#8203;:
> 
> Hello\,
> 
> > On Sun Mar 06 10&#8203;:20&#8203;:12 2011\, corion@&#8203;cpan\.org wrote&#8203;:
> >> To get a Perlbug number\, I'm resending the patch appended
> >> to this mail\.
>  >> \[\.\.\.\]
> >>
> > Is there any reason I should not apply this?
> >
> 
> I'm unaware of any such reason\. I couldn't get a reply either for or 
> against it from p5p at the time\, and also unfortunately didn't get the 
> deprecation message into 5\.14\.
> 
> Still I think it should get applied\, as the two CPAN modules affected by 
> it have patches\, and I think it fixes more latent bugs than it causes\.

I’ve just applied it as 0e21945\. I tweaked it a bit\, as your version was
allowing $$ assignment up until the first fetch \(just moved the
SvREADONLY\_on to gv\.c\)\.

Thank you\!

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
